### ISSUE TITLE
New Rule: disallow use of an unbound class method in a callback

### ISSUE THREAD

**Issue by markelliot** [palantir/tslint#259]
I'd like to detect and prevent patterns such as:

``` ts
    public foo() { /* body */ }
    public bar() {
        registerCallback(this.foo); // should be registerCallback(() => { this.foo(); });
                                    // or registerCallback(this.foo.bind(this));
    }
```

**Comment by JKillian** [comment]
:+1:

**Comment by myitcv** [comment]
:+1: we had a discussion about this just last week.

Suspect we need #680 in order to do this properly?

**Comment by adidahiya** [comment]
it seems possible to do syntactically, without type information, but it's not an optimal solution:

1. in one pass over the file AST, keep track of all the class methods for each class in the file.
2. in a second pass, look for `CallExpression` nodes within instance methods and scan their arguments. if one of the arguments references another instance method without using the fat arrow or `.bind` syntax, report a failure

^ this would probably fail with class hierarchies, though, because you wouldn't know which `super...` values are methods and which are members if the parent class is declared outside this file.

so yeah, looks like full type information would be needed to do this properly

**Comment by adidahiya** [comment]
Note: this rule should apply to JSX as well:

```ts
// bad
public render() {
    return <form onSubmit={this.handleSubmit}>
}

private handleSubmit(e: React.SyntheticEvent) {
    ...
}
```
